File: app/src/main/java/com/marcelorcorrea/falae/fragment/TabPagerFragment.java
package com.marcelorcorrea.falae.fragment;

import android.content.Context;
import android.net.Uri;
import android.os.Bundle;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;
import android.support.v4.view.ViewPager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.marcelorcorrea.falae.R;
import com.marcelorcorrea.falae.model.SpreadSheet;
import com.marcelorcorrea.falae.model.User;

public class TabPagerFragment extends Fragment implements SpreadSheetFragment.OnFragmentInteractionListener, UserInfoFragment.OnFragmentInteractionListener {

    private static final String USER_PARAM = "userParam";
    private User user;

    private OnFragmentInteractionListener mListener;
    private ViewPager viewPager;

    public TabPagerFragment() {
    }

    public static TabPagerFragment newInstance(User user) {
        TabPagerFragment fragment = new TabPagerFragment();
        Bundle args = new Bundle();
        args.putParcelable(USER_PARAM, user);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            user = getArguments().getParcelable(USER_PARAM);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_tab_pager, container, false);
        viewPager = (ViewPager) view.findViewById(R.id.viewpager);
        viewPager.setAdapter(new CustomFragmentPagerAdapter(getChildFragmentManager(), getActivity()));
        viewPager.setOffscreenPageLimit(2);

        // Give the TabLayout the ViewPager
        TabLayout tabLayout = (TabLayout) view.findViewById(R.id.layout_tabs);
        tabLayout.setupWithViewPager(viewPager);
        return view;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnFragmentInteractionListener) {
            mListener = (OnFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    @Override
    public void displayActivity(SpreadSheet spreadSheet) {
        mListener.displayActivity(spreadSheet);

    }

    @Override
    public void onFragmentInteraction(Uri uri) {

    }

    public interface OnFragmentInteractionListener {
        void displayActivity(SpreadSheet spreadSheet);
    }

    public class CustomFragmentPagerAdapter extends FragmentStatePagerAdapter {
        final int TAB_COUNT = 2;
        private String tabTitles[] = new String[]{getResources().getString(R.string.spreadsheets), getResources().getString(R.string.user_info)};
        private Context context;

        public CustomFragmentPagerAdapter(FragmentManager fm, Context context) {
            super(fm);
            this.context = context;
        }

        public Fragment getItem(int position) {
            Fragment fragment;
            if (position == 0) {
                fragment = SpreadSheetFragment.newInstance(user);
            } else {
                fragment = UserInfoFragment.newInstance(user);
            }
            return fragment;
        }

        @Override
        public CharSequence getPageTitle(int position) {
            return tabTitles[position];
        }

        @Override
        public int getCount() {
            return TAB_COUNT;
        }
    }
}

File: app/src/main/java/com/marcelorcorrea/falae/fragment/TabPagerFragment.kt
package com.marcelorcorrea.falae.fragment

import android.content.Context
import android.net.Uri
import android.os.Bundle
import android.support.design.widget.TabLayout
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v4.app.FragmentStatePagerAdapter
import android.support.v4.view.ViewPager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.marcelorcorrea.falae.R
import com.marcelorcorrea.falae.model.SpreadSheet
import com.marcelorcorrea.falae.model.User

class TabPagerFragment : Fragment(), SpreadSheetFragment.OnFragmentInteractionListener, UserInfoFragment.OnFragmentInteractionListener {

    private var user: User? = null

    private lateinit var mListener: OnFragmentInteractionListener
    private lateinit var viewPager: ViewPager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            user = arguments.getParcelable(USER_PARAM)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_tab_pager, container, false)
        viewPager = view.findViewById(R.id.viewpager) as ViewPager
        viewPager.adapter = CustomFragmentPagerAdapter(childFragmentManager, activity)
        viewPager.offscreenPageLimit = 2

        // Give the TabLayout the ViewPager
        val tabLayout = view.findViewById(R.id.layout_tabs) as TabLayout
        tabLayout.setupWithViewPager(viewPager)
        return view
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (context is OnFragmentInteractionListener) {
            mListener = context
        } else {
            throw RuntimeException(context!!.toString() + " must implement OnFragmentInteractionListener")
        }
    }

    override fun onDetach() {
        super.onDetach()
    }

    override fun displayActivity(spreadSheet: SpreadSheet) {
        mListener.displayActivity(spreadSheet)
    }

    override fun removeUser(user: User) {
        mListener.removeUser(user)
    }

    override fun onFragmentInteraction(uri: Uri) {

    }

    interface OnFragmentInteractionListener {
        fun displayActivity(spreadSheet: SpreadSheet)
        fun removeUser(user: User)
    }

    inner class CustomFragmentPagerAdapter(fm: FragmentManager, private val context: Context) : FragmentStatePagerAdapter(fm) {
        private val TAB_COUNT = 2
        private val tabTitles = arrayOf(resources.getString(R.string.spreadsheets), resources.getString(R.string.user_info))

        override fun getItem(position: Int): Fragment {
            return if (position == 0) {
                SpreadSheetFragment.newInstance(user!!)
            } else {
                UserInfoFragment.newInstance(user!!)
            }
        }

        override fun getPageTitle(position: Int): CharSequence = tabTitles[position]

        override fun getCount(): Int = TAB_COUNT
    }

    companion object {

        private val USER_PARAM = "userParam"

        fun newInstance(user: User?): TabPagerFragment {
            val fragment = TabPagerFragment()
            val args = Bundle()
            args.putParcelable(USER_PARAM, user)
            fragment.arguments = args
            return fragment
        }
    }
}
